vm_trace.c

```diff
@@ -1617,197 +1621,289 @@ Init_vm_trace(void)
 size_t
-rb_vm_memsize_postponed_job_buffer(void)
+rb_vm_memsize_postponed_job_queues(void)
 {
-    return sizeof(rb_postponed_job_t) * MAX_POSTPONED_JOB;
+    rb_postponed_job_queues_t *pjq = GET_VM()->postponed_job_queues;
+    size_t sz = sizeof(*pjq);
+    if (pjq->workqueue_accumulating) {
+        sz += st_memsize(pjq->workqueue_accumulating);
+    }
+    if (pjq->workqueue_draining) {
+        sz += st_memsize(pjq->workqueue_draining);
+    }
+    return sz;
 }
 
-void
-Init_vm_postponed_job(void)
+
+static inline bool
+pjob_disable_gc(void)
 {
-    rb_vm_t *vm = GET_VM();
-    vm->postponed_job_buffer = ALLOC_N(rb_postponed_job_t, MAX_POSTPONED_JOB);
-    vm->postponed_job_index = 0;
-    /* workqueue is initialized when VM locks are initialized */
+    if (!ruby_thread_has_gvl_p()) {
+        return false;
+    }
+    return !RB_TEST(rb_gc_disable_no_rest());
 }
 
-enum postponed_job_register_result {
-    PJRR_SUCCESS     = 0,
-    PJRR_FULL        = 1,
-    PJRR_INTERRUPTED = 2
-};
-
-/* Async-signal-safe */
-static enum postponed_job_register_result
-postponed_job_register(rb_execution_context_t *ec, rb_vm_t *vm,
-                       unsigned int flags, rb_postponed_job_func_t func, void *data, rb_atomic_t max, rb_atomic_t expected_index)
+static inline void
+pjob_reenable_gc(bool was_enabled)
 {
-    rb_postponed_job_t *pjob;
+    if (!was_enabled) {
+        return;
+    }
+    rb_gc_enable();
+}
 
-    if (expected_index >= max) return PJRR_FULL; /* failed */
+rb_postponed_job_handle
+rb_postponed_job_preregister(rb_postponed_job_func_t func, void *data)
+{
+    /* The doc comments say that this function should be called under the GVL, because
+     * that is actually required to get the guarantee that "if a given (func, data) pair
+     * was already pre-registered, this method will return the same handle instance".
+     *
+     * However, the actual implementation here is called without the GVL, from inside
+     * rb_postponed_job_register, to support that legacy interface. In the presence
+     * of concurrent calls to both _preregister and _register functions on the same
+     * func, however, the data may get mixed up between them. */
 
-    if (ATOMIC_CAS(vm->postponed_job_index, expected_index, expected_index+1) == expected_index) {
-        pjob = &vm->postponed_job_buffer[expected_index];
-    }
-    else {
-        return PJRR_INTERRUPTED;
+    rb_postponed_job_queues_t *pjq = GET_VM()->postponed_job_queues;
+    for (unsigned int i = 0; i < PJOB_PREREG_TABLE_SIZE; i++) {
+        /* Needs to be atomic: must synchronize with RUBY_ATOMIC_PTR_CAS below */
+        rb_postponed_job_func_t existing_func = (rb_postponed_job_func_t)RUBY_ATOMIC_PTR_LOAD(pjq->prereg_table[i].func);
+        /* Needs to be atomic: must synchronize with RUBY_ATOMIC_PTR_EXCHANGE below */
+        void *existing_data = RUBY_ATOMIC_PTR_LOAD(pjq->prereg_table[i].data);
+cas_retry:
+        if (existing_func == func && existing_data == data) {
```

Comment:
should we check existing_data too?

Reply:
That is what I documented in the header file:

```
* Pre-registers a (func, data) pair in Ruby's postponed job preregistration table,
 * returning an opaque handle which can be used to trigger the job later. This function
 * must be called under the GVL (generally, it will be called during the initialization
 * routine of an extension).
 *
 * If the given (func,data) pair was already pre-registered, this method will return
 * the same handle instance as was previously returned.
 ```

Honestly I wrote the documentation first, and this "felt" like the most obvious API design, to think in terms of `(func, data)` pairs. Otherwise, you'd need to write something like "If this function is called multiple times with the same func and different data, the data argument is ignored and it returns the same handle which was pre-registered the first time `func` was registered", which just seems a little wordier.

Reply:
> "If this function is called multiple times with the same func and different data, the data argument is ignored and it returns the same handle which was pre-registered the first time func was registered", which just seems a little wordier.

Yes. This is what I assume. Otherwise, 64 (or 32) entries are not enough. Now we confirmed that the number of `func` is enough small. If we allow multiple `data` we can not register finalizers for example. This is not acceptable.

Reply:
(for finalizer (and other important `func`s) will be reserved by preregister, though)

Reply:
that's right - the internal usages of this are pre-registered during early boot, before any user code or other extensions can run - so extensions can't possibly starve out Ruby's own need for postponed job slots.

Reply:
Ruby 3.3 RC1 will be released soon so:

* you will change the spec.
* merge this commit and change the spec after that by me.

which do you want?

I'm sorry to make non-frequent responses.

Reply:
So about this code example... (note. I purposely changed `i * 1000` to `i + 1000` to make something clearer)

```
  // 1st iteration
  for (i=0; i<100; i++) rb_postponed_job_register(a_func, i);
  // 2nd iteration
  for (i=0; i<100; i++) rb_postponed_job_register(a_func, i + 1000);
```

I definitely understand how my change as it currently is will break this. What I don't think I understand is how you _want_ this to behave.

Right now, on master, this code will cause `a_func` to be called 100 times with the numbers 0 ... 100, and then 100 more times with the numbers 1000...1100.

What do you want the behaviour of this to be?

1. make `a_func` be called 100 times with argument `0`, and then, later on,  100 more times with argument `0`
2. make `a_func` be called one time with argument `0` , and then, later on, one more time with argument `0`
3. make `a_func` be called one time with argument `0`, and then, later on, one more time with argument `1000`
4. make `a_func` be called ~32 times with the numbers 0..32 as an argument, and then never again after that? (that's what my current patch is going to do)

I think if we remove ` && existing_data == data` from the check on line 1759, the behaviour will be option 2.

I guess what worries me is that anything actually using `rb_postponed_job_register` is going to probably wind up doing something very different than the author originally intended with any of these options. 

> Ruby 3.3 RC1 will be released soon 

To step back a bit, maybe what we should do is:

* Wait for Ruby 3.3 to be released,
* Early next year, merge the circular, lock-free ringbuffer implementation I originally proposed. I definitely agree it was too complicated to merge in December a month before Ruby 3.3, but maybe it's OK to merge in January?
* Gives us plenty of time to stabilise it and work out the bugs, or revert it if it's becoming a problem.
* Because the ringbuffer implementation keeps the existing semantics of both `rb_postponed_job_register` and `rb_postponed_job_register_one`, we can consider it for backport into a point release once we're happy with it.

What do you think? Should we leave this until after 3.3? As far as I can tell I'm the only one who ran into the signal-safety crash in the first place, and this bug will only affect people using profiling tools in production, so it's probably not worth rushing it into 3.3 I think.

Reply:
> make a_func be called one time with argument 0, and then, later on, one more time with argument 1000

Let decide to choose option 3. I proposed it.

>  I guess what worries me is that anything actually using rb_postponed_job_register is going to probably wind up doing something very different than the author originally intended with any of these options.

Yes the behavior will be changed. However the size of queue is implementation details and this change is 1000 -> 1. It is not a semantic change but a parameter change (yes, it change the behavior).

> What do you think? Should we leave this until after 3.3? As far as I can tell I'm the only one who ran into the signal-safety crash in the first place, and this bug will only affect people using profiling tools in production, so it's probably not worth rushing it into 3.3 I think.

I want to introduce my proposal on Ruby 3.3.0. If someone blamed the change, we can revert fix on Ruby 3.3.1.

Reply:
> make a_func be called one time with argument 0, and then, later on, one more time with argument 1000

Sorry. Let's make clear:

* On iteration 1, register `a_func` with 0.
* one call `a_func(0)` on safe-place
* On iteration 2, register `a_func` with 1000.
* one call `a_func(1000)` on safe-place
* ...

(`one more time with argument 1000` is not clear, so I make it clear)

Reply:
@ko1 OK, I _think_ I have implemented that functionality correctly and pushed the change up. It still needs a couple more test cases I think, but otherwise does this look how you wanted it to?

Also: Do we need to open a discussion on Redmine about this kind of change, do you think?

Reply:
Thank you for the long discussion. I'll merge it.

One thing I find we need `flag` parameter for future change but I'll fix it later.

Making redmine ticket is good idea. Could you summary the fixed spec and reasons?

Reply:
I added some trivial comments so please refer to refine if they make sense.